File: LeetCodeChallenges/TwoSum/TwoSum.cpp
class Solution {
    
public:
    vector<int> twoSumSlow(vector<int>& nums, int target) {
        for(int i = 0; i < nums.size(); i++) {
            for(int j = 1; j < nums.size(); j++) {
                if(i != j && (nums[i] + nums[j]) == target)  {
                    return vector{i, j};
                }
            }
        }
        // Indicates that no solution was found
        assert(false && "No solution was found");
    }
    
    vector<int> twoSumFast(vector<int>& nums, int target) {
        unordered_map<int, unordered_set<int>> numCache;
        for(int i = 0; i < nums.size(); i++) {
            // automatically calls set default constructor if nums[i] is not in numCache
            auto & indexSet = numCache[nums[i]];
            indexSet.insert(i);
        }
        
        for(int j = 0; j < nums.size(); j++) {
            int required = target - nums[j];
            if(numCache.find(required) != numCache.end()) {
                auto & indexSet = numCache[required];
                for(int index: indexSet) {
                    if(index != j) {
                        return vector{j, index};
                    }
                }
            }
        }
        
        // Indicates that no solution was found
        assert(false && "No solution was found");
    }
    
    vector<int> twoSum(vector<int>& nums, int target) {
        return twoSumFast(nums, target);
    }
};

File: LeetCodeChallenges/TwoSum/TwoSum.py
class Solution:
    def twoSumFast(self, nums, target):
        numberCache = defaultdict(set)
        for i, val in enumerate(nums):
            numberCache[val].add(i)
        
        for j, val in enumerate(nums):
            if (indexSet := numberCache.get(target - val)) is not None:
                for entry in indexSet:
                    if entry != j:
                        return j, entry
    
    def twoSumSlow(self, nums, target):
        for i, val1 in enumerate(nums):
            for j, val2 in enumerate(nums):
                if i != j and (val1 + val2) == target:
                    return [i, j]
    
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        return self.twoSumFast(nums, target)